### DISCUSSION TITLE
How use SCC with actixweb like Cache for backend

### DISCUSSION THREAD

**Discussion by andreysneg** [wvwwvwwv/scalable-concurrent-containers#142] [General]
I want to make a cache for all users (threads) of ACTIXWEB, so that they can read and write to scc:Hasmap.

This code doesn't work. What would be more correct? ARC is it needed here ? 

```
pub async fn main() -> std::io::Result<()> {

    // Cache create
    let dt:Arc<HashMap<String,String>> = Arc::new(HashMap::default());

    HttpServer::new(|| {
        App::new()
            .app_data(web::Data::new(Arc::clone(&dt)))
            .default_service(web::to(default))
    })
        .bind("ansa:81")?
        .run()
        .await
}
pub async fn default( req: HttpRequest,  cache: web::Data<Arc<Cache>>) -> impl Responder {

    let cache = (*cache).get_value();

}
```

**Comment by wvwwvwwv** [comment]
No, I believe an `Arc` is not needed here and you can pass `&dt` to the server, but not 100% sure; I'm not familiar with Actix Web.

Unfortunately, I have no idea what you want to achieve with the piece of code - `dt` is always empty so getting value from it doesn't make any sense to me. Can you elaborate on the code please?

=> One minor thing - I also don't know what `get_value()` is, but you must call async methods of `SCC`, otherwise there is a possibility of deadlock.

**Comment by andreysneg** [comment]
For example, in Tokyo i create 1000 async threads and 1 css:hashmap.  I want all threads to use 1 css:hashmap

Do I need to insert it into the ARC ?

**Reply by wvwwvwwv** [reply]
As long as the Rust compiler doesn't complain about the lifetime of it (e.g., in this case, there is a chance that `dt` is destroyed before the server ends, or `dt` might be moved), you don't have to use `Arc`.

**Comment by andreysneg** [comment]
This code is working.
dt:Date is an analog of Arc in actixweb. 
Without date/arch it’s not clear how to use it here yet...

The idea is to create REDIS analogue for web backend.  100x faster.


```
pub async fn main() -> std::io::Result<()> {

    let mut dt:Data<HashMap<i32,i32>> = Data::new(HashMap::new());
    
    HttpServer::new(move || {
        App::new()
            .app_data(dt.clone())
            .default_service(web::to(default))
       })
        .bind("ansa:81")?
        .run()
        .await
}

pub async fn default(req: HttpRequest, cache: Data<HashMap<i32,i32>> ) -> impl Responder {
    
        cache.insert(1,2); 	

	HttpResponse::Ok().body("")
}
```

**Reply by wvwwvwwv** [reply]
The code itself looks good enough to me except that I recommend you replace `cache.insert(1, 2)` with `cache.insert_async(1, 2).await`. If `Actix Web` provides `Data`, then this will be the idiomatic way of passing a value to the server I think, but again, I don't know much about `Actix Web`..
